br/pkg/task/restore_data.go

```diff
@@ -131,14 +131,14 @@ func RunResolveKvData(c context.Context, g glue.Glue, cmdName string, cfg *Resto
 	}
 
 	log.Debug("total tikv", zap.Int("total", numBackupStore), zap.String("progress file", cfg.ProgressFile))
-	// progress = read meta + send recovery + iterate tikv + resolve kv data.
-	progress := g.StartProgress(ctx, cmdName, int64(numBackupStore*4), !cfg.LogProgress)
-	go progressFileWriterRoutine(ctx, progress, int64(numBackupStore*4), cfg.ProgressFile)
+	// progress = read meta + send recovery + iterate tikv + prepare region + flashback.
+	progress := g.StartProgress(ctx, cmdName, int64(numBackupStore*3+2), !cfg.LogProgress)
+	go progressFileWriterRoutine(ctx, progress, int64(numBackupStore*3+2), cfg.ProgressFile)
 
 	// restore tikv data from a snapshot volume
 	var totalRegions int
 
-	totalRegions, err = restore.RecoverData(ctx, resolveTs, allStores, mgr, progress)
+	totalRegions, err = restore.RecoverData(ctx, resolveTs, allStores, mgr, progress, restoreTS, cfg.Concurrency)
```

Comment:
it seems `restoreTs` > `resolveTs`. is it ok to stop gc at `restoreTs`?

Reply:
keep safe point in restoreTs is safe since GC never work before the TiDB node startup. The code may need a refactoring to remove the safepointkeeper service, however, I would like to remove it after a thorough test.